### ISSUE TITLE
on Python 3, make setup.py install the modules that have been ported to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6040]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6040|
|Type|enhancement|
|Created|2012-10-03 11:44:50Z|
|Branch|https://github.com/twisted/twisted/tree/setup-py3-6040-4|

`setup.py` on Python 3 should install all modules that have been ported.

Unfortunately tests probably won't run out of an installed twisted package, since a few of the test modules rely on modules that haven't been ported yet, but that should resolve itself over time.

<details><summary>Searchable metadata</summary>

```
trac-id__6040 6040
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_setup_py3_6040_4 branches/setup-py3-6040-4
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1349264690000000 1349264690000000
changetime__1350403009000000 1350403009000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

I love issues that resolve themselves.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [36073]) Branching to 'setup-py3-6040'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [36103]) Branching to 'setup-py3-6040-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [36159]) Branching to 'setup-py3-6040-3'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @itamarst<br>@exarkun set status to `new`|
|-|-|

Eh, well, this still doesn't produce a distribution you can run the tests on, because at least `twisted.internet.posixbase` is still missing from the ported list.  However, I guess that situation will be remedied whenever posixbase does get added to that list, without any further changes needed to this new setup script.  So perhaps it's worth just landing this now (of course this could be irrational optimism and it may turn out that even once all the modules are available there is something else missing...).

Anyway, I'll let you take a look and see what you think.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst commented|
|-|-|

Note that even with posixbase, we'll have issues running all of tests out of the installed version. Specifically, some of ported trial tests depend on importing some other trial modules that haven't yet been ported.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

I propose that we add a third category of module, then.  Not-really-ported-but-needed-by-ported.  Otherwise we're never going to be able to run tests on the package we're generating and we won't have a very good idea of whether the package actually works or not.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst set owner to @exarkun|
|-|-|

Thanks for figuring this out.

  1. News file; maybe `setup3.py` is actually worth mentioning there?
  1. future imports in `setup3.py`.
  1. The moved `getAllScripts` looks like it needs a test? You could just open a ticket for that, since it's unchanged in Python 2 and unused in Python 3.
  1. Having a third category makes sense, yes. You should add all modules necessary for tests to run. Hopefully just trial stuff (I'm going to merge posixbase next).
  1. I was actually going to add abstract.py with a new ticket that also covered `twisted.test.test_abstract`; you can move it to not-really-ported list, or leave it where it is, up to you.

Fix, then merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [36194]) Branching to 'setup-py3-6040-4'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

Thanks!

> News file; maybe setup3.py is actually worth mentioning there?

r36191

> future imports in setup3.py.

r36192

> The moved getAllScripts looks like it needs a test? You could just open a ticket for that, since it's unchanged in Python 2 and unused in Python 3.

Filed [#6137](https://github.com/twisted/twisted/issues/6137)

> Having a third category makes sense, yes. You should add all modules necessary for tests to run. Hopefully just trial stuff (I'm going to merge posixbase next).
> I was actually going to add abstract.py with a new ticket that also covered twisted.test.test_abstract; you can move it to not-really-ported list, or leave it where it is, up to you.

r36193, r36196, and r36197

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [36198]) Merge setup-py3-6040-4

Author: exarkun
Reviewer: itamarst
Fixes: [#6040](https://github.com/twisted/twisted/issues/6040)

Add `setup3.py` script which can package and install a subset of Twisted on Python 3.
